contracts/GroupHealth.sol

```diff
@@ -0,0 +1,297 @@
+                .getTotalStCeloVotesForStrategy(group);
+
+            return (
+                manager.toCelo(supposedStCeloInActiveGroup + supposedstCeloInSpecificGroupStrategy),
+                realCelo
+            );
+        }
+
+        revert GroupNotEligible(group);
+    }
+
+    /**
+     * @notice Rebalances Celo between groups that have incorrect Celo-stCelo ratio.
+     * FromGroup is required to have more Celo than it should and ToGroup needs
+     * to have less Celo than it should.
+     * @param fromGroup The from group.
+     * @param toGroup The to group.
+     */
+    function rebalance(address fromGroup, address toGroup)
+        public
+        view
+        returns (
+            address[] memory fromGroups,
+            address[] memory toGroups,
+            uint256[] memory fromVotes,
+            uint256[] memory toVotes
+        )
+    {
+        uint256 expectedFromCelo;
+        uint256 realFromCelo;
+
+        if (
+            !defaultStrategy.groupsContain(toGroup) &&
+            !specificGroupStrategy.isSpecificGroupStrategy(toGroup)
+        ) {
+            // rebalancinch to deprecated/non-existant group is not allowed
+            revert InvalidToGroup(toGroup);
+        }
+
+        if (fromGroup == address(0)) {
+            revert InvalidFromGroup(fromGroup);
+        }
+
+        (expectedFromCelo, realFromCelo) = getExpectedAndRealCeloForGroup(fromGroup);
+
+        if (realFromCelo <= expectedFromCelo) {
+            // fromGroup needs to have more Celo than it should
+            revert RebalanceNoExtraCelo(fromGroup, realFromCelo, expectedFromCelo);
+        }
+
+        uint256 expectedToCelo;
+        uint256 realToCelo;
+
+        (expectedToCelo, realToCelo) = getExpectedAndRealCeloForGroup(toGroup);
+
+        if (realToCelo >= expectedToCelo) {
+            // toGroup needs to have less Celo than it should
+            revert RebalanceEnoughCelo(toGroup, realToCelo, expectedToCelo);
+        }
+
+        fromGroups = new address[](1);
+        toGroups = new address[](1);
+        fromVotes = new uint256[](1);
+        toVotes = new uint256[](1);
+
+        fromGroups[0] = fromGroup;
+        fromVotes[0] = Math.min(realFromCelo - expectedFromCelo, expectedToCelo - realToCelo);
+
+        toGroups[0] = toGroup;
+        toVotes[0] = fromVotes[0];
+    }
+
+    /**
+     * @notice Checks if a group member is elected.
+     * @param groupMember The member of the group to check election status for.
+     * @return Whether or not the group member is elected.
+     */
+    function isGroupMemberElected(address groupMember) private view returns (bool) {
+        IElection election = getElection();
+
+        address[] memory electedValidatorSigners = election.electValidatorSigners();
+
+        for (uint256 i = 0; i < electedValidatorSigners.length; i++) {
+            if (electedValidatorSigners[i] == groupMember) {
+                return true;
+            }
+        }
```

Comment:
Wait. Is this going to run through the whole election process and iterate through all (100+) elected validators on each iteration of the loop that checks if the group has elected members? That's A LOT of gas.

Reply:
Looks like the way to get the currently elected validators is using the `validatorSignerAddressFromCurrentSet` precompile. Would need the caller to provide the index of the validator to avoid the 100+ iteration loop over the set.

Reply:
Agreed about the gas usage, but wouldn't using `validatorSignerAddressFromCurrentSet(index)` require that you already know if a group is elected?

Reply:
Yes. Which top level Contract calls end up needing this check? If every deposit, and especially if ERC20 transfers, require this, then yeah, that would be a no-go at this point. But I think a redesign of when we check group eligibility could enable this. The idea:
* keep a mapping in SpecificGroupStrategy of which groups have been marked as eligible.
* In transfers/deposits/etc, just trust this mapping as valid (i.e. no health check on deposit to a specific group).
* Have functions like `markGroupEligible`/`markGroupIneligible` that can be called to update the mapping. Provide the appropriate indexes to these functions to allow quick onchain verification.

Reply:
I see your point. This was originally intended to be used when adding or deprecating groups, so not often used. 

To build on your idea, `specificStrategy.allowStrategy()` already checks group eligibility before adding the group to the set. So during deposits and strategy changes, using `specificGroupStrategy.isSpecificGroupStrategy()` should prevent the excessive use of `isValidGroup`

Reply:
Group Health now has cache.

Unfortunately with this design comes a downside. Since for updating the group health status it is necessary to provide correct index. It is possible to invalidate group by just providing incorrect index on purpose. Because of this I implemented mechanism that doesn't allow to invalidate group in same epoch in which this group was proclaimed as valid.

Reply:
Refactored as discussed offline